### PR TITLE
Add `use-t-throws-async-well` rule

### LINKED ISSUE #274: `t.throwsAsync()` and `t.notThrowsAsync()` should be awaited on
[Issuehunt badges](https://issuehunt.io/r/avajs/eslint-plugin-ava/issues/274)

When you use the `t.throwsAsync()` and `t.notThrowsAsync()` assertion, you must await the promise they return. If the test function completes before the assertions do the test will fail.

We should add a rule to ensure users invoke these assertions with the `await` keyword. I'm not sure whether it should be recommended — you *could* assign the promise to a variable and await it later, but I don't think that's typical when you write tests.

See also the discussion in https://github.com/avajs/ava/issues/2245.

@sindresorhus?

---

> *Note: This issue has a bounty, so it's expected that you are an experienced programmer and that you give it your best effort if you intend to tackle this. Don't forget, if applicable, to add tests, docs (double-check for typos). And don't be sloppy. Review your own diff multiple times and try to find ways to improve and simplify your code. Instead of asking too many questions, present solutions. The point of an issue bounty is to reduce my workload, not give me more. Include a 🦄 in your PR description to indicate that you've read this. Thanks for helping out 🙌* - @sindresorhus

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Backers (Total: $60.00)

- [issuehunt issuehunt](https://issuehunt.io/u/issuehunt) ($60.00)
### Submitted pull Requests
- [#278 Add `use-t-throws-async-well` rule](https://issuehunt.io/r/avajs/eslint-plugin-ava/pull/278)
---

### Tips

- Checkout the [Issuehunt explorer](https://issuehunt.io/r/avajs/eslint-plugin-ava/) to discover more funded issues.
- Need some help from other developers? [Add your repositories](https://issuehunt.io/r/new) on IssueHunt to raise funds.
---
IssueHunt has been backed by the following sponsors. [Become a sponsor](https://issuehunt.io/membership/members)
</details>

### PR DESCRIPTION
**Pull request by stroncium**
Fixes #274

Ensure that `t.throwsAsync()` and `t.notThrowsAsync()` aren't called without await(or some other action using returned promise).

:unicorn:

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Referenced issues

This pull request has been submitted to:
- [#274: `t.throwsAsync()` and `t.notThrowsAsync()` should be awaited on](https://issuehunt.io/repos/49722492/issues/274)
---

IssueHunt has been backed by the following sponsors. [Become a sponsor](https://issuehunt.io/membership/members)
</details>

### CODE REVIEW

File: docs/rules/use-t-throws-async-well.md [hunk 1]
```diff
@@ -0,0 +1,28 @@
+# Ensure that `t.throwsAsync()` and `t.notThrowsAsync()` aren't called without await
```
Comment (sindresorhus) [inline docs/rules/use-t-throws-async-well.md]:
```suggestion
# Ensure that `t.throwsAsync()` and `t.notThrowsAsync()` are awaited
```

File: docs/rules/use-t-throws-async-well.md [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+```
+
+
+## Pass
+
+```js
+import test from 'ava';
+
+test('main', t => {
+	await t.throwsAsync(() => {});
```
Comment (sindresorhus) [inline docs/rules/use-t-throws-async-well.md]: This is not valid. It should be a promise passed to `t.throwsAsync()`.

**Comment by sindresorhus** [comment]
Can you add an auto-fixer? It would just add `await` in front if it's missing, but only if it's inside an `async` function.

**Comment by stroncium** [comment]
@sindresorhus done

**Comment by sindresorhus** [comment]
Looks good. IssueHunt is down, so remind me another day and I'll reward it.

**Comment by stroncium** [comment]
@sindresorhus ping =D

**Comment by stroncium** [comment]
@sindresorhus BTW, any info on how issuehunt is doing? Seems dying out to me.

**Comment by sindresorhus** [comment]
No updates unfortunately: https://spectrum.chat/issuehunt/general/error-when-accessing-an-issue-page~edd3be59-9228-472f-ae32-ce888e56bc1c

You could try emailing them at kazz@boostio.co

**Comment by stroncium** [comment]
Wait, I can access anything including issue pages.

What I meant is dying out as a project. It was the most promising one to put opensource on commercial rails.

**Pull request merged**